internal/cli/command/login/oidc.go

```diff
@@ -0,0 +1,343 @@
+			http.Error(w, errMsg+": "+r.FormValue("error_description"), http.StatusBadRequest)
+			return
+		}
+		code := r.FormValue("code")
+		if code == "" {
+			http.Error(w, fmt.Sprintf("no code in request: %q", r.Form), http.StatusBadRequest)
+			return
+		}
+		if state := r.FormValue("state"); state != exampleAppState {
+			http.Error(w, fmt.Sprintf("expected state %q got %q", exampleAppState, state), http.StatusBadRequest)
+			return
+		}
+		token, err = oauth2Config.Exchange(ctx, code)
+	case http.MethodPost:
+		// Form request from frontend to refresh a token.
+		refresh := r.FormValue("refresh_token")
+		if refresh == "" {
+			http.Error(w, fmt.Sprintf("no refresh_token in request: %q", r.Form), http.StatusBadRequest)
+			return
+		}
+		t := &oauth2.Token{
+			RefreshToken: refresh,
+			Expiry:       time.Now().Add(-time.Hour),
+		}
+		token, err = oauth2Config.TokenSource(ctx, t).Token()
+	default:
+		http.Error(w, fmt.Sprintf("method not implemented: %s", r.Method), http.StatusBadRequest)
+		return
+	}
+
+	if err != nil {
+		http.Error(w, fmt.Sprintf("failed to get token: %v", err), http.StatusInternalServerError)
+		return
+	}
+
+	rawIDToken, ok := token.Extra("id_token").(string)
+	if !ok {
+		http.Error(w, "no id_token in token response", http.StatusInternalServerError)
+		return
+	}
+
+	idToken, err := a.verifier.Verify(r.Context(), rawIDToken)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Failed to verify ID token: %v", err), http.StatusInternalServerError)
+		return
+	}
+
+	var claims json.RawMessage
+	idToken.Claims(&claims)
+
+	buff := new(bytes.Buffer)
+	json.Indent(buff, []byte(claims), "", "  ")
+
+	err = a.requestTokenFromPipeline(rawIDToken)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("failed to request Pipeline token: %v", err), http.StatusInternalServerError)
+		return
+	}
+
+	renderClosingTemplate(w)
+
+	log.Info("successfully logged in")
+}
+
+func (a *app) requestTokenFromPipeline(rawIDToken string) error {
+	pipelineURL, err := url.Parse(a.pipelineBasePath)
+	if err != nil {
+		return err
+	}
+
+	pipelineURL.Path = "/auth/dex/callback"
+
+	reqBody := bytes.NewBuffer(nil)
+	writer := multipart.NewWriter(reqBody)
+	writer.WriteField("id_token", rawIDToken)
+	writer.Close()
+
+	client := &http.Client{
+		CheckRedirect: func(req *http.Request, via []*http.Request) error {
+			return http.ErrUseLastResponse
+		},
+	}
+
+	resp, err := client.Post(pipelineURL.String(), writer.FormDataContentType(), reqBody)
+	if err != nil {
+		return err
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode < 200 || resp.StatusCode >= 400 {
+		body, _ := ioutil.ReadAll(resp.Body)
+		return fmt.Errorf("request returned: %s", string(body))
+	}
+
+	for _, cookie := range resp.Cookies() {
+		if cookie.Name == "user_sess" {
+			a.banzaiCli.Context().SetToken(cookie.Value)
+			return nil
+		}
+	}
+
+	return fmt.Errorf("failed to find user_sess cookie in Pipeline response")
+}
+
+func open(url string) error {
```

Comment:
could use [github.com/pkg/browser](https://godoc.org/github.com/pkg/browser#OpenURL) instead (not sure if it's worth an extra dep)